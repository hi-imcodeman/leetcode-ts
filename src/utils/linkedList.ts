
//Definition for singly-linked list.
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

export const arrayToLinkedList = (nums:number[]):ListNode | null => {
    if(!nums.length)
        return null
    const listNode:ListNode = new ListNode(nums[0])
        let currNode = listNode
        let i=1
        while(i<nums.length){
            const node = new ListNode(nums[i])
            currNode.next = node
            currNode = currNode.next
            i++
        }
        return listNode
}

export const linkedListToArray = (listNode:ListNode|null):number[] => {
    if(!listNode)
        return []
    const data:number[]=[]
    let currNode = listNode
    data.push(currNode.val)
    while(currNode.next){
        currNode = currNode.next
        data.push(currNode.val)
    }
    return data
}

