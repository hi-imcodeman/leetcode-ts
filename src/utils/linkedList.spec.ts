import { arrayToLinkedList, linkedListToArray } from './linkedList'

describe('Linked List', () => {
    test('arrayToLinkedList',()=>{
        const LL = arrayToLinkedList([1,2,3,4,5,6])
        expect(LL).toMatchSnapshot()
    })
    test('linkedListToArray',()=>{
        const LL = arrayToLinkedList([1,2,3,4,5,6])
        const arr = linkedListToArray(LL)
        expect(arr).toEqual([1,2,3,4,5,6])
    })
})