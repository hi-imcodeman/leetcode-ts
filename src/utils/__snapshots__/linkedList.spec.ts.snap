// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linked List arrayToLinkedList 1`] = `
ListNode {
  "next": ListNode {
    "next": ListNode {
      "next": ListNode {
        "next": ListNode {
          "next": ListNode {
            "next": null,
            "val": 6,
          },
          "val": 5,
        },
        "val": 4,
      },
      "val": 3,
    },
    "val": 2,
  },
  "val": 1,
}
`;
